group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2427,25 +2471,259 @@ private boolean hasMemberSubscriptionChanged(
+        }
+
+        // If the number of subscribed regexes is different from the number of resolved ones,
+        // we must trigger a refresh.
+        requireRefresh |= subscribedRegularExpressions.size() != group.numResolvedRegularExpressions();
+
+        // If the last time that regexes were refreshed is older than the refresh interval,
+        // we must trigger a refresh.
+        requireRefresh |= time.milliseconds() >
+            group.lastResolvedRegularExpressionRefreshTimeMs() + consumerGroupRegularExpressionRefreshInternalMs;
+
+        if (requireRefresh) {
+            Set<String> regexes = subscribedRegularExpressions.keySet();
+            executor.schedule(
+                key,
+                () -> refreshRegularExpressions(groupId, log, time, metadataImage, regexes),
+                (result, exception) -> handleRegularExpressionsResult(groupId, result, exception)
+            );
+        }
+
+        return bumpGroupEpoch;
+    }
+
+    /**
+     * Resolves the provided regular expressions. Note that this static method is executed
+     * as an asynchronous task in the executor. Hence, it should not access any state from
+     * the manager.
+     *
+     * @param groupId   The group id.
+     * @param log       The log instance.
+     * @param time      The time instance.
+     * @param image     The metadata image to use for listing the topics.
+     * @param regexes   The list of regular expressions that must be resolved.
+     * @return The list of resolved regular expressions.
+     */
+    private static Map<String, ResolvedRegularExpression> refreshRegularExpressions(
+        String groupId,
+        Logger log,
+        Time time,
+        MetadataImage image,
+        Set<String> regexes
+    ) {
+        log.debug("[GroupId {}] Refreshing regular expressions: {}", groupId, regexes);
+        Map<String, Set<String>> resolvedRegexes = new HashMap<>(regexes.size());
+        List<Pattern> compiledRegexes = new ArrayList<>(regexes.size());
+        for (String regex : regexes) {
+            resolvedRegexes.put(regex, new HashSet<>());
+            try {
+                compiledRegexes.add(Pattern.compile(regex));
+            } catch (PatternSyntaxException ex) {
+                // This should not happen because the regular expressions are validated
+                // when received from the members. If for some reasons, it would
+                // happen, we log it and ignore it.
+                log.error("[GroupId {}] Couldn't parse regular expression '{}' due to `{}`. Ignoring it.",
+                    groupId, regex, ex.getDescription());
+            }
+        }
+
+        for (String topicName : image.topics().topicsByName().keySet()) {
+            for (Pattern regex : compiledRegexes) {
+                if (regex.matcher(topicName).matches()) {
+                    resolvedRegexes.get(regex.pattern()).add(topicName);
+                }
+            }
+        }
```

Comment:
fundamentally, we have to iterate on all topics to match a given regex. i am not sure how we could improve this. do you have suggestions? the operation is basically proportional to the number of topics. i’ll try to get some numbers.

Reply:
i pushed a small jmh benchmark for this method. here are the results:

```
Benchmark                     (regexCount)  (topicCount)  Mode  Cnt      Score     Error  Units
RegexResolutionBenchmark.run             1         10000  avgt    5      7.126 ±   0.751  ms/op
RegexResolutionBenchmark.run             1        100000  avgt    5     82.415 ±   1.493  ms/op
RegexResolutionBenchmark.run             1       1000000  avgt    5    941.900 ±  35.111  ms/op
RegexResolutionBenchmark.run            10         10000  avgt    5     54.309 ±   4.544  ms/op
RegexResolutionBenchmark.run            10        100000  avgt    5    592.577 ±  29.491  ms/op
RegexResolutionBenchmark.run            10       1000000  avgt    5   7891.792 ± 222.076  ms/op
RegexResolutionBenchmark.run           100         10000  avgt    5    132.687 ±   7.707  ms/op
RegexResolutionBenchmark.run           100        100000  avgt    5   1441.310 ±  32.055  ms/op
RegexResolutionBenchmark.run           100       1000000  avgt    5  15439.901 ± 720.625  ms/op
JMH benchmarks done
```

Reply:
I agree the amount of work to do a single run is inevitable and the benchmark results look acceptable (although i'm not sure how much the length of the topic names affect the performance)

The concern I had was the amount of duplicate work we do across long periods of time, specifically when topics do not change && subscribedRegex stays the same.

My understanding is that the refresh interval is only required to catch any changes in the topics image. What if in GMM#onNewMetadataImage, if there were any created/deleted topics we require a refresh for all groups. This wouldn't remove all duplicate work, but will reduce it substantially if we assume `topics created/deleted rate <<< once every 5 minutes`

Wdyt?

Reply:
Yeah, I was considering this too. My reasoning was that in large clusters, topics creation/deletion rate is more frequent than the 5 minutes refresh interval. The second thing to consider is that we would need to refresh all regexes after a failover because we could miss topics created/deleted while offline. I suppose that both approaches are valid depending on the cluster. Let me think about it.

Reply:
Here are the results with three words:
```
Benchmark                     (regexCount)  (topicCount)  Mode  Cnt      Score     Error  Units
RegexResolutionBenchmark.run             1         10000  avgt    5      9.566 ±   0.360  ms/op
RegexResolutionBenchmark.run             1        100000  avgt    5    109.216 ±   1.943  ms/op
RegexResolutionBenchmark.run             1       1000000  avgt    5   1191.775 ±  22.478  ms/op
RegexResolutionBenchmark.run            10         10000  avgt    5     74.200 ±   5.143  ms/op
RegexResolutionBenchmark.run            10        100000  avgt    5    971.574 ±  10.051  ms/op
RegexResolutionBenchmark.run            10       1000000  avgt    5   9243.389 ±  71.810  ms/op
RegexResolutionBenchmark.run           100         10000  avgt    5    182.088 ±   2.064  ms/op
RegexResolutionBenchmark.run           100        100000  avgt    5   1936.796 ±  33.055  ms/op
RegexResolutionBenchmark.run           100       1000000  avgt    5  20297.554 ± 807.382  ms/op
JMH benchmarks done
```

Reply:
> topics creation/deletion rate is more frequent than the 5 minutes refresh interval

That was something i didn't know. I think we can go with the PR implementation, add some metrics for visibility, and revisit this. An additional word seems to bring roughly +30% increase which seems quite large